src/azure-cli-core/azure/cli/core/azlogging.py

```diff
@@ -177,6 +177,10 @@ def end_cmd_metadata_logging(self, exit_code):
         if self.command_metadata_logger:
             self.command_metadata_logger.info("exit code: %s", exit_code)
 
+            for handler in self.command_metadata_logger.handlers[:]:
+                handler.close()
+                self.command_metadata_logger.removeHandler(handler)
```

Comment:
>  self.command_metadata_logger.removeHandler(handler) [](start = 14, length = 53)

remove elements in the array loop may cause unexpected result

Reply:
it's a copy from slice `[:]`, won't have side affect.

Reply:
Will this get cleaned up if the `az` process returns an error? I seem to remember the call to `end_cmd_metadata_logging` being skipped if an exception was raised when I was testing as it was only called on a successful path rather than from a `finally` block.

Reply:
@elpollouk If you were using `az`, that would go with code in `__main__.py` 

https://github.com/Azure/azure-cli/blob/1bab7985899f267b208cfe6048fb253485357c4c/src/azure-cli/azure/cli/__main__.py#L48-L73

whenever an exception is raised, the final exception on top of exception stack would be `SystemExit` because there is a hook:
https://github.com/Azure/azure-cli/blob/4770a437fb9fbef119b50248bc9853c9e3fb9855/src/azure-cli-core/azure/cli/core/parser.py#L148-L157

That's my understanding, if anything missing/wrong, could you please give a detailed report?